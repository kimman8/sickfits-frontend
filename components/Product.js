import Link from 'next/link';
import styled from 'styled-components';
import formatMoney from '../lib/formatMoney';
import ItemStyles from './styles/ItemStyles';
import PriceTag from './styles/PriceTag';

const KimsTitle = styled.h3`
  margin: 0 1rem;
  text-align: center;
  position: relative;
  transform: skew(-5deg) rotate(-1deg);
  margin-top: -3rem;
  text-shadow: 2px 2px 0 rgba(0, 0, 0, 0.1);

  a {
    background: var(--red);
    padding: 0 1rem;
    display: inline;
    line-height: 1.3;
    color: white;
    font-size: 4rem;
  }
`;

const PriceStyle = styled.h4`
  transform: rotate(3deg);
  background: var(--red);
  position: absolute;
  color: white;
  font-size: 3rem;
  margin: 0;
  text-align: center;
  padding: 0 1rem;
  display: inline-block;
  font-weight: 600;
  line-height: 1;
  padding: 5px;
  top: -3px;
  right: -3px;
`;

export default function Product({ product }) {
  return (
    <ItemStyles>
      <img
        src={product?.photo?.image?.publicUrlTransformed}
        alt={product.name}
      />

      <KimsTitle>
        <Link href={`/product/${product.id}`}>{product.name}</Link>
      </KimsTitle>
      <PriceStyle>{formatMoney(product.price)}</PriceStyle>
      <p>{product.description}</p>
    </ItemStyles>
  );
}
