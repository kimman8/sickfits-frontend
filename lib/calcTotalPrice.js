export default function calcTotalPrice(cart) {
  return cart.reduce((tally, cartItem) => {
    if (!cartItem.product) return tally; // products can be deleted but they could still be in y our cart
    return tally + cartItem.product.price * cartItem.quantity;
  }, 0);
}
