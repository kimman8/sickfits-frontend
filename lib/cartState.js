import React, { createContext, useContext, useState } from 'react';

const LocalStateContext = createContext();
const LocalStateProvider = LocalStateContext.Provider;

function CartStateProvider({ children }) {
  // this is our own custom provider! we will store data(state) and functionality (updates) in here and anyone can access it via the consumer!

  const [cartOpen, setCartOpen] = useState(false);
  function toggleCart() {
    setCartOpen(!cartOpen);
  }

  function closeCart() {
    setCartOpen(false);
  }

  function openCart() {
    setCartOpen(true);
  }
  return (
    <LocalStateProvider
      value={{ cartOpen, setCartOpen, toggleCart, closeCart, openCart }}
    >
      {children}
    </LocalStateProvider>
  );
}
// make a custom hook for accessing the cart local state
function useCart() {
  // we use a consumer here to access the local state
  const all = useContext(LocalStateContext);
  return all;
}
export { CartStateProvider, useCart };
