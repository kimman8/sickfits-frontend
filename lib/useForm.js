import { useState } from 'react';

export default function useForm(initial = {}) {
  // create a state object for our inputs
  const [inputs, setInputs] = useState(initial);
  function handleChange(e) {
    let { type, value, name } = e.target;
    if (type === 'number') {
      value = parseInt(value);
    }
    if (type === 'file') {
      [value] = e.target.files;
    }
    setInputs({
      // copy existing state
      ...inputs,
      [name]: value,
    });
  }

  function resetForm() {
    setInputs(initial);
  }

  function clearForm() {
    const blankState = Object.fromEntries(
      Object.entries(inputs).map(([key, value]) => [key, ''])
    );
    setInputs(blankState);
  }
  return {
    inputs,
    handleChange,
    resetForm,
    clearForm,
  };
}
