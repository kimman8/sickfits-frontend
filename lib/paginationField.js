import { PAGINATION_QUERY } from '../components/Pagination';

export default function paginationField() {
  return {
    keyArgs: false, // tells apollo we will take care of everything
    read(existing = [], { args, cache }) {
      console.log({ existing, args, cache });
      const { skip, first } = args;
      // read the number of items on the page from the cache
      const data = cache.readQuery({ query: PAGINATION_QUERY });
      console.log({ data });
      const count = data?._allProductsMeta?.count;
      const page = skip / first + 1;
      const pages = Math.ceil(count / first);

      // check if we have existing items
      const items = existing.slice(skip, skip + first).filter((x) => x);

      // if there are items, just return them from the cache, and we dont need to go to the network
      if (items.length && items.length !== first && page === pages) {
        return items;
      }
      if (items.length !== first) {
        return false;
      } // fallback to network if the 2 above if statements dont work
      if (items.length) {
        return items;
      }
      return false;
    },
    merge(existing, incoming, { args }) {
      const { skip, first } = args;
      // this runs when the apollo client comes back from the network with our product
      console.log(`mergin items from network ${incoming.length} `);
      console.log({ incoming, existing });
      const merged = existing ? existing.slice(0) : [];
      for (let i = skip; i < skip + incoming.length; ++i) {
        merged[i] = incoming[i - skip];
      }
      console.log(merged);
      // finally we return the merged items from the cache
      return merged;
    },
  };
}
