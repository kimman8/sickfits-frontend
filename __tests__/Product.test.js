import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import Product from '../components/Product';
import { fakeItem } from '../lib/testUtils';
import { CartStateProvider } from '../lib/cartState';

const product = fakeItem();
describe('<Product/>', () => {
  it('renders out the price tag and title', () => {
    const { container, debug } = render(
      <CartStateProvider>
        <MockedProvider>
          <Product product={product} />
        </MockedProvider>
      </CartStateProvider>
    );
    const priceTag = screen.getByText('$50');
    expect(priceTag).toBeInTheDocument();
    const link = container.querySelector('a');
    debug(link);
    expect(link).toHaveAttribute('href', '/product/abc123');
    expect(link).toHaveTextContent(product.name);
  });
  it('renders and matches the snapshot', () => {
    const { container, debug } = render(
      <CartStateProvider>
        <MockedProvider>
          <Product product={product} />
        </MockedProvider>
      </CartStateProvider>
    );
    expect(container).toMatchSnapshot();
  });
  it('renders image properly', () => {
    const { container, debug } = render(
      <CartStateProvider>
        <MockedProvider>
          <Product product={product} />
        </MockedProvider>
      </CartStateProvider>
    );
    // grab da image
    const img = screen.getByAltText(product.name);
    expect(img).toBeInTheDocument();
  });
});
