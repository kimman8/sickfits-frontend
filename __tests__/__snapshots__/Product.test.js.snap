// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Product/> renders and matches the snapshot 1`] = `
<div>
  <div
    class="ItemStyles-sc-16pk14u-0 iBOgQI"
  >
    <img
      alt="dogs are best"
      src="dog.jpg"
    />
    <h3
      class="Title-sc-16nq74k-0 kDFcWM"
    >
      <a
        href="/product/abc123"
      >
        dogs are best
      </a>
    </h3>
    <span
      class="PriceTag-nwbk6t-0 iFDMLq"
    >
      $50
    </span>
    <p>
      dogs
    </p>
    <div
      class="buttonList"
    >
      <a
        href="/update?id=abc123"
      >
        Edit ✏️
      </a>
      <button
        type="button"
      >
        Add
         to Cart 🛒
      </button>
      <button
        type="button"
      >
        Delete 🗑
      </button>
    </div>
  </div>
</div>
`;
